src/faebryk/libs/util.py

```diff
@@ -854,3 +855,77 @@ def zip_exhaust(*args):
 
 def join_if_non_empty(sep: str, *args):
     return sep.join(s for arg in args if (s := str(arg)))
+
+
+class FuncSet[T](collections.abc.Set[T]):
+    """
+    A set by pre-processing the objects with the hasher function.
+    """
+
+    def __init__(
+        self, data: Sequence[T] | None = None, hasher: Callable[[T], Hashable] = id
+    ):
+        self._hasher = hasher
+        self._deref = {hasher(d): d for d in data} if data else {}
```

Comment:
has to be a multidict (=defaultdict(list))
because hash doesn't imply unique

__contains__ would then need to check if obj in self._deref.get(obj, [])

Reply:
This should behave like a `set` in python. I don't think we should create a back-door to stick in multiple de-referenced entries for a given hash

Reply:
sets in python are hashsets and thus can handle collisions:

```python
class A:
    def __init__(self, value: int) -> None:
        self.value = value

    def __hash__(self) -> int:
        return 42

    def __eq__(self, __value: object) -> bool:
        assert isinstance(__value, A)
        return self.value == __value.value

    def __repr__(self) -> str:
        return f"A({self.value})"


a1 = A(10)
a2 = A(20)

seta = {a1, a2}

print(hash(a1) == hash(a2))  # True
print(a1 == a2)  # False
print(seta)  # {A(10), A(20)}
```